File: vbnet/reporting/03_export_targets.vb
' 03_export_targets -- VB.NET port of examples\Vb6\reporting\03_export_targets.bas
Imports System
Imports System.IO
Imports System.Runtime.InteropServices
Imports LumasPdfSdk

Module mod03_export_targets
    Private Const PDF_DEMO_KEY As String = "LUMAS-LumasReportExamples-DD5D40E0"
    Private Const RPT_DEMO_KEY As String = "LRPT2-TFIwMgECAAIAAgIAAQADBAD___9_BAEAAAUBAAAGCAAAAAAAAAAAAAcQAEx1bWFzUnB0RXhhbXBsZXM.sneCg6LX_J59AcqzUZPx1WqQ7jGQZyji4z0WMm_vIoycNySw12vGVoKV76gk3-yQtx3LfZZepGxoDdQAgf5MAA"

    Private mPdf As IntPtr
    Private mEng As IntPtr

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Function TrimNull(ByVal s As String) As String
        If s Is Nothing Then Return ""
        Dim p As Integer = s.IndexOf(ChrW(0))
        If p >= 0 Then Return s.Substring(0, p) Else Return s
    End Function

    Sub WriteTextFile(ByVal path As String, ByVal content As String)
        File.WriteAllText(path, content)
    End Sub

    Function GetRptErr(ByVal eng As IntPtr, ByRef info As TRptErrorInfoC) As Boolean
        Dim sz As Integer = Marshal.SizeOf(GetType(TRptErrorInfoC))
        Dim p As IntPtr = Marshal.AllocHGlobal(sz)
        Try
            Dim ok As Boolean = LumasPdf.rptGetLastError(eng, p)
            If ok Then info = CType(Marshal.PtrToStructure(p, GetType(TRptErrorInfoC)), TRptErrorInfoC)
            Return ok
        Finally
            Marshal.FreeHGlobal(p)
        End Try
    End Function

    Sub DumpRptError(ByVal eng As IntPtr)
        Dim info As TRptErrorInfoC
        If GetRptErr(eng, info) Then
            If info.Code <> 0 Then
                Console.WriteLine("  ! rpt error " & info.Code & " [" & TrimNull(info.Module_) & _
                    "] at " & TrimNull(info.Location) & ": " & TrimNull(info.Msg))
            End If
        End If
    End Sub

    Function BootEngine() As Boolean
        mPdf = LumasPdf.pdfNewPDF()
        If mPdf = IntPtr.Zero Then
            Console.WriteLine("pdfNewPDF failed")
            Return False
        End If
        LumasPdf.pdfSetLicenseKey(mPdf, PDF_DEMO_KEY)
        LumasPdf.rptSetRptLicenseKeyA(mPdf, RPT_DEMO_KEY)
        mEng = LumasPdf.rptCreateEngineA(mPdf, Nothing)
        If mEng = IntPtr.Zero Then
            Console.WriteLine("rptCreateEngine failed:")
            DumpRptError(IntPtr.Zero)
            Return False
        End If
        Return True
    End Function

    Function FileSizeOf(ByVal path As String) As Long
        Try
            Return New FileInfo(path).Length
        Catch
            Return -1
        End Try
    End Function

    Sub Main()
        Dim Targets() As Integer = {LumasPdfConsts.RPT_EXP_PDF, LumasPdfConsts.RPT_EXP_HTML, LumasPdfConsts.RPT_EXP_CSV, _
            LumasPdfConsts.RPT_EXP_JSON, LumasPdfConsts.RPT_EXP_XML, LumasPdfConsts.RPT_EXP_TEXT, LumasPdfConsts.RPT_EXP_SVG, _
            LumasPdfConsts.RPT_EXP_XLSX, LumasPdfConsts.RPT_EXP_PNG, LumasPdfConsts.RPT_EXP_BMP}
        Dim Exts() As String = {"pdf", "html", "csv", "json", "xml", "txt", "svg", "xlsx", "png", "bmp"}

        If Not BootEngine() Then Return

        Dim Csv As String = AppPath() & "\03_data.csv"
        Dim Lrpt As String = AppPath() & "\03_report.lrpt"
        Dim CsvData As String = ""
        CsvData = CsvData & "product,qty,price" & vbLf
        CsvData = CsvData & "Widget,4,9.95" & vbLf
        CsvData = CsvData & "Gadget,2,19.50" & vbLf
        CsvData = CsvData & "Sprocket,7,3.25" & vbLf
        WriteTextFile(Csv, CsvData)
        Dim Xml As String = ""
        Xml = Xml & "<?xml version=""1.0"" encoding=""UTF-8""?>" & vbLf
        Xml = Xml & "<report name=""ExportDemo"" tagLangVersion=""1"">" & vbLf
        Xml = Xml & " <page width=""210"" height=""297"" marginLeft=""15"" marginTop=""15"" marginRight=""15"" marginBottom=""15""/>" & vbLf
        Xml = Xml & " <datasources><datasource alias=""d"" provider=""csv"" conn=""" & Csv & """/></datasources>" & vbLf
        Xml = Xml & " <bands>" & vbLf
        Xml = Xml & "  <band kind=""reportheader"" name=""rh"" height=""14"">" & vbLf
        Xml = Xml & "   <text name=""ttl"" x=""0"" y=""0"" w=""180"" h=""8"" fontSize=""16"" hAlign=""center"">Order Lines</text>" & vbLf
        Xml = Xml & "  </band>" & vbLf
        Xml = Xml & "  <band kind=""detail"" name=""det"" height=""7"" data=""d"">" & vbLf
        Xml = Xml & "   <text name=""p"" x=""0""   y=""0"" w=""90"" h=""6"" fontSize=""10"" wordWrap=""0"">{{d.product}}</text>" & vbLf
        Xml = Xml & "   <text name=""q"" x=""90""  y=""0"" w=""30"" h=""6"" fontSize=""10"" hAlign=""right"" wordWrap=""0"">{{d.qty}}</text>" & vbLf
        Xml = Xml & "   <text name=""r"" x=""120"" y=""0"" w=""60"" h=""6"" fontSize=""10"" hAlign=""right"" wordWrap=""0"">{{d.price}}</text>" & vbLf
        Xml = Xml & "  </band>" & vbLf
        Xml = Xml & " </bands>" & vbLf
        Xml = Xml & "</report>" & vbLf
        WriteTextFile(Lrpt, Xml)

        Dim Job As IntPtr = LumasPdf.rptOpenReportA(mEng, Lrpt)
        If Job = IntPtr.Zero Then
            Console.WriteLine("open failed")
            DumpRptError(mEng)
            GoTo Cleanup
        End If
        If Not LumasPdf.rptRender(Job) Then
            Console.WriteLine("render failed")
            DumpRptError(mEng)
            LumasPdf.rptCloseReport(Job)
            GoTo Cleanup
        End If
        Console.WriteLine("rendered " & LumasPdf.rptGetPageCount(Job) & " page(s)")
        Console.WriteLine("== Exporting to all targets ==")
        Dim i As Integer
        For i = 0 To 9
            Dim OutFile As String = AppPath() & "\03_out." & Exts(i)
            If LumasPdf.rptExportA(Job, Targets(i), OutFile) AndAlso File.Exists(OutFile) Then
                Console.WriteLine("  [" & Exts(i) & "] id=" & Targets(i) & "  OK  " & FileSizeOf(OutFile) & " bytes")
            Else
                Console.WriteLine("  [" & Exts(i) & "] id=" & Targets(i) & "  FAILED")
                DumpRptError(mEng)
            End If
        Next
        LumasPdf.rptCloseReport(Job)
Cleanup:
        LumasPdf.rptDeleteEngine(mEng)
        LumasPdf.pdfDeletePDF(mPdf)
    End Sub
End Module

File: python/reporting/03_export_targets.py
import sys, os, ctypes
import os, sys
# the wrapper lives at <root>/wrappers/python (source checkout) or beside
# the example tree (shipped package) -- find it without a hard-coded path
_d = os.path.dirname(os.path.abspath(__file__))
for _ in range(6):
    for _c in (os.path.join(_d, 'wrappers', 'python'), _d):
        if os.path.isfile(os.path.join(_c, 'lumaspdf.py')):
            sys.path.insert(0, _c)
            break
    else:
        _d = os.path.dirname(_d)
        continue
    break
import lumaspdf as L

# ===========================================================================
#  Python (ctypes) port of examples\Vb6\reporting\03_export_targets.bas
# ===========================================================================

PDF_DEMO_KEY = "LUMAS-LumasReportExamples-DD5D40E0"
RPT_DEMO_KEY = "LRPT2-TFIwMgECAAIAAgIAAQADBAD___9_BAEAAAUBAAAGCAAAAAAAAAAAAAcQAEx1bWFzUnB0RXhhbXBsZXM.sneCg6LX_J59AcqzUZPx1WqQ7jGQZyji4z0WMm_vIoycNySw12vGVoKV76gk3-yQtx3LfZZepGxoDdQAgf5MAA"

HERE = os.path.dirname(os.path.abspath(__file__))


def trim_null(s):
    return s.split('\0')[0]


def write_text(path, content):
    with open(path, 'w', encoding='utf-8') as f:
        f.write(content)


def dump_rpt_error(eng):
    info = L.TRptErrorInfoC()
    if L.rptGetLastError(eng, ctypes.byref(info)) != 0:
        if info.Code != 0:
            print("  ! rpt error %d [%s] at %s: %s" % (
                info.Code,
                trim_null(info.Module_.decode('latin-1')),
                trim_null(info.Location.decode('latin-1')),
                trim_null(info.Msg.decode('latin-1'))))


def boot_engine():
    pdf = L.pdfNewPDF()
    if not pdf:
        print("pdfNewPDF failed")
        return None, None
    L.pdfSetLicenseKey(pdf, PDF_DEMO_KEY.encode('latin-1'))
    L.rptSetRptLicenseKeyA(pdf, RPT_DEMO_KEY.encode('latin-1'))
    eng = L.rptCreateEngineA(pdf, None)
    if not eng:
        print("rptCreateEngine failed:")
        dump_rpt_error(0)
        return None, pdf
    return eng, pdf


def main():
    targets = [
        (L.RPT_EXP_PDF, "pdf"), (L.RPT_EXP_HTML, "html"), (L.RPT_EXP_CSV, "csv"),
        (L.RPT_EXP_JSON, "json"), (L.RPT_EXP_XML, "xml"), (L.RPT_EXP_TEXT, "txt"),
        (L.RPT_EXP_SVG, "svg"), (L.RPT_EXP_XLSX, "xlsx"), (L.RPT_EXP_PNG, "png"),
        (L.RPT_EXP_BMP, "bmp"),
    ]

    eng, pdf = boot_engine()
    if not eng:
        return

    csv = os.path.join(HERE, "03_data.csv")
    lrpt = os.path.join(HERE, "03_report.lrpt")
    csv_data = (
        "product,qty,price\n"
        "Widget,4,9.95\n"
        "Gadget,2,19.50\n"
        "Sprocket,7,3.25\n"
    )
    write_text(csv, csv_data)
    xml = (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<report name="ExportDemo" tagLangVersion="1">\n'
        ' <page width="210" height="297" marginLeft="15" marginTop="15" marginRight="15" marginBottom="15"/>\n'
        ' <datasources><datasource alias="d" provider="csv" conn="' + csv + '"/></datasources>\n'
        ' <bands>\n'
        '  <band kind="reportheader" name="rh" height="14">\n'
        '   <text name="ttl" x="0" y="0" w="180" h="8" fontSize="16" hAlign="center">Order Lines</text>\n'
        '  </band>\n'
        '  <band kind="detail" name="det" height="7" data="d">\n'
        '   <text name="p" x="0"   y="0" w="90" h="6" fontSize="10" wordWrap="0">{{d.product}}</text>\n'
        '   <text name="q" x="90"  y="0" w="30" h="6" fontSize="10" hAlign="right" wordWrap="0">{{d.qty}}</text>\n'
        '   <text name="r" x="120" y="0" w="60" h="6" fontSize="10" hAlign="right" wordWrap="0">{{d.price}}</text>\n'
        '  </band>\n'
        ' </bands>\n'
        '</report>\n'
    )
    write_text(lrpt, xml)

    try:
        job = L.rptOpenReportA(eng, lrpt.encode('latin-1'))
        if not job:
            print("open failed")
            dump_rpt_error(eng)
            return
        if L.rptRender(job) == 0:
            print("render failed")
            dump_rpt_error(eng)
            L.rptCloseReport(job)
            return
        print("rendered %d page(s)" % L.rptGetPageCount(job))
        print("== Exporting to all targets ==")
        for tid, ext in targets:
            out_file = os.path.join(HERE, "03_out." + ext)
            if L.rptExportA(job, tid, out_file.encode('latin-1')) != 0 and os.path.exists(out_file):
                print("  [%s] id=%d  OK  %d bytes" % (ext, tid, os.path.getsize(out_file)))
            else:
                print("  [%s] id=%d  FAILED" % (ext, tid))
                dump_rpt_error(eng)
        L.rptCloseReport(job)
    finally:
        L.rptDeleteEngine(eng)
        L.pdfDeletePDF(pdf)


if __name__ == "__main__":
    main()
